### PR TITLE
ITDEV-35098 Reset ACLs

### PR DESCRIPTION
**Pull request by ris-bmulligan**
Testing instructions:

1. Build locally from `feature/ITDEV-35098-reset-acls` branch
2. Connect to VPN
3. Login, create allocations as needed and navigate to View All Allocations page
4. Choose an allocation from the list of available allocations.
5. Near the bottom of the Update Allocation form, click "Reset ACLs" with "Include sub-allocations" checkbox checked as needed.

### CODE REVIEW

File: coldfront/plugins/qumulo/views/update_allocation_view.py [hunk 1]
```diff
@@ -61,7 +77,55 @@ def get_form_kwargs(self):
...
+            has_linkage = False
+        return has_linkage
+
+
+    def _reset_acls(self):
+        # bmulligan (20240903): "retry" and "timeout" are intended to be
+        # arbitrarily high values.  It would be good if the application could
+        # know or learn what they should be.  Testing has shown that the DEV
+        # infrastructure can process a directory tree of about 90,500 items in
+        # 62 minutes.
```
Comment (peaches7083) [inline coldfront/plugins/qumulo/views/update_allocation_view.py:108]: Not sure where you're using `logger` inside the `_reset_acls` scope

File: coldfront/plugins/qumulo/views/update_allocation_view.py [hunk 2]
```diff
@@ -61,7 +77,55 @@ def get_form_kwargs(self):
...
+        except AllocationLinkage.DoesNotExist:
+            has_linkage = False
+        return has_linkage
+
+
+    def _reset_acls(self):
+        # bmulligan (20240903): "retry" and "timeout" are intended to be
+        # arbitrarily high values.  It would be good if the application could
+        # know or learn what they should be.  Testing has shown that the DEV
+        # infrastructure can process a directory tree of about 90,500 items in
```
Comment (peaches7083) [inline coldfront/plugins/qumulo/views/update_allocation_view.py:107]: RE: knowing how long - SysEng is setting up DataDog.  That *should* (if incorporated with the app) be able to show us a graph of completion times for tasks by type.

Meanwhile, we could just log start/end times inside the tasks running on qcluster.

File: coldfront/plugins/qumulo/views/update_allocation_view.py (hunk 1 above)
Comment (ris-bmulligan) [reply coldfront/plugins/qumulo/views/update_allocation_view.py:108]: That was just for debugging.

File: coldfront/templates/common/authorized_navbar.html [hunk 3]
```diff
@@ -35,7 +35,7 @@
           <a class="nav-link" href="{% url 'qumulo:allocation' %}">Create Allocation</a>
         </li>
         <li id="navbar-view-allocation" class="nav-item">
-          <a class="nav-link" href="{% url 'qumulo:allocation-table-list' %}">View All Allocation</a>
+          <a class="nav-link" href="{% url 'qumulo:allocation-table-list' %}">View All Allocations</a>
```
Comment (melonistic) [inline coldfront/templates/common/authorized_navbar.html:38]: @ris-bmulligan thanks for fixing the spelling. A minor change that makes the app presentable.

File: coldfront/plugins/qumulo/views/update_allocation_view.py [hunk 4]
```diff
@@ -90,38 +147,47 @@ def form_valid(
...
 
-        # RW and RO users are not handled via an AllocationChangeRequest
         access_keys = ["rw", "ro"]
         for key in access_keys:
             access_users = form_data[key + "_users"]
             self.set_access_users(key, access_users, allocation)
 
-        return super().form_valid(form=form, parent_allocation=parent_allocation)
+    def form_valid(self, form: UpdateAllocationForm):
+        if 'reset_acls' in self.request.POST:
```
Comment (melonistic) [inline coldfront/plugins/qumulo/views/update_allocation_view.py:186]: @ris-bmulligan why is the method a POST? I thought it should be PUT or DELETE?

File: coldfront/plugins/qumulo/hooks.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            width=80
+        )
+    email_body += (
+        '\n\n'
+        'If you need anything further, then please contact User Support.'
+        '\n\n'
+        'Thanks,\n\n'
+        'Coldfront\n\n'
+    )
+    if EMAIL_ENABLED:
```
Comment (melonistic) [inline coldfront/plugins/qumulo/hooks.py:38]: @ris-bmulligan in the future, we should consider an email interceptor.

File: coldfront/plugins/qumulo/hooks.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        send_email(
+            subject,
+            (
+                'Greetings,\n\n'
+                f'{email_body}'
+            ),
+            'ris@wustl.edu',
```
Comment (melonistic) [inline coldfront/plugins/qumulo/hooks.py:45]: @ris-bmulligan perhaps, the sender should be moved to a configuration.

File: coldfront/plugins/qumulo/hooks.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            [recipient]
+        )
+    else:
+        log_file = ENV.str(
+            'ACL_RESET_MSG_LOG',
+            default='/tmp/e-mail_notifications.log'
+        )
+        with open(log_file, 'a') as emnl:
+            emnl.write(
+                (
```
Comment (melonistic) [inline coldfront/plugins/qumulo/hooks.py:55]: @ris-bmulligan move to a template file. This way the template can be reused whether is sent by email or written to a file.

File: coldfront/plugins/qumulo/hooks.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+def acl_reset_complete_hook(task_object):
+    task_duration = '{:.2f}'.format(task_object.time_taken())
+    allocation_name = task_object.args[1].get_attribute('storage_name')
+    recipient = task_object.args[0]
+    subject = f'{allocation_name} ACL Reset Task'
+    email_body = ''
+    if task_object.success:
+        email_body += textwrap.fill(
+            (
+                'Your request to perform an ACL Reset '
```
Comment (melonistic) [inline coldfront/plugins/qumulo/hooks.py:15]: @ris-bmulligan templetize

File: coldfront/plugins/qumulo/hooks.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                f'(task "{task_object.name}") on allocation "{allocation_name}" '
+                f'completed successfully after {task_duration} seconds.'
+            ),
+            width=80
+        )
+    else:
+        email_body += textwrap.fill(
+            (
```
Comment (melonistic) [inline coldfront/plugins/qumulo/hooks.py:23]: @ris-bmulligan templetize

File: coldfront/plugins/qumulo/hooks.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import textwrap
+
+from coldfront.config.env import ENV
+from coldfront.core.utils.mail import EMAIL_ENABLED, send_email
+
+def acl_reset_complete_hook(task_object):
+    task_duration = '{:.2f}'.format(task_object.time_taken())
```
Comment (melonistic) [inline coldfront/plugins/qumulo/hooks.py:7]: @ris-bmulligan I would refactor out into at least two methods:

1. calculate what goes into the variables
2. another handling the send email or write to file responsibilties.

File: coldfront/plugins/qumulo/tasks.py [hunk 11]
```diff
@@ -163,3 +167,158 @@ def __validate_results(quota_usages, logger) -> bool:
...
+        self.ro_group = AclAllocations.get_allocation_rwro_group_name(
+            access_allocations,
+            'ro'
+        )
+        self.reset_exclude_paths = []
+        if self.is_allocation_root:
+            self.sub_allocations = []
+            links = None
+            try:
+                links = AllocationLinkage.objects.get(parent=allocation)
```
[the remark this answers is not in the document]
Comment (melonistic) [reply coldfront/plugins/qumulo/tasks.py:197]: however, did you meant to say filter(...).first()

File: coldfront/plugins/qumulo/tasks.py [hunk 12]
```diff
@@ -163,3 +167,157 @@ def __validate_results(quota_usages, logger) -> bool:
...
+                    )
+                    self.sub_allocations.append(child)
+        else:
+            self._setup_qumulo_api()
+            self.parent_aces = AclAllocations.get_sub_allocation_parent_aces(
+                allocation,
+                self.qumulo_api
+            )
+
+    # <debugging functions>
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tasks.py:215]: @ris-bmulligan in future, these may be moved to a mixin. On the other hand, we should also agree on how to use different log levels. I would leave this alone; for now however, it should be a good conversation for setting development logging expectations.

File: coldfront/plugins/qumulo/views/update_allocation_view.py [hunk 13]
```diff
@@ -11,21 +14,30 @@
...
 from coldfront.plugins.qumulo.utils.acl_allocations import AclAllocations
 from coldfront.plugins.qumulo.utils.active_directory_api import ActiveDirectoryAPI
 
 
+logger = logging.getLogger(__name__)
+
 class UpdateAllocationView(AllocationView):
     form_class = UpdateAllocationForm
-    template_name = "allocation.html"
+    template_name = "update_allocation.html"
```
Comment (melonistic) [inline coldfront/plugins/qumulo/views/update_allocation_view.py:34]: @ris-bmulligan is this related to the reset acls?

File: coldfront/plugins/qumulo/views/update_allocation_view.py [hunk 14]
```diff
@@ -75,9 +87,54 @@ def get_form_kwargs(self):
...
+                    .objects \
+                    .get(pk=self.kwargs.get("allocation_id")) \
+                    .get_attribute(name='storage_name')
+        return (
+            f'ACL reset initiated for {name}.  '
+            'An e-mail notification will be sent when the task completes.'
+        )
+
+    def _allocation_linkage_exists(self):
+        has_linkage = True
```
Comment (melonistic) [inline coldfront/plugins/qumulo/views/update_allocation_view.py:101]: @ris-bmulligan would you consider getting a result set and check if it is empty? This way the try-except would not be necessary since all it's required is to determined whether or not some records exists.

File: coldfront/plugins/qumulo/tasks.py [hunk 15]
```diff
@@ -163,3 +167,157 @@ def __validate_results(quota_usages, logger) -> bool:
...
+        global logger
+        if not self.debug:
+            return None
+        logger.warn(msg)
+
+    def __log_acl_reset(self, path):
+        self.__log_acl(f'Resetting "directory content" ACLs for path: {path}')
+    # </debugging functions>
+
+    def _setup_qumulo_api(self):
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tasks.py:226]: @ris-bmulligan what are the use cases when the QUMULO API is not required? Perhaps, this is telling that this class has too many responsibilities. A sub-class with more responsibilities or a decorator could separate the concerns.

File: coldfront/plugins/qumulo/views/update_allocation_view.py (hunk 4 above)
Comment (ris-bmulligan) [reply coldfront/plugins/qumulo/views/update_allocation_view.py:186]: The submit button does a POST.

File: coldfront/plugins/qumulo/hooks.py (hunk 5 above)
Comment (ris-bmulligan) [reply coldfront/plugins/qumulo/hooks.py:38]: At the time that part was put together, I don't think e-mail was enabled--and it remains that way for local or development environments.  More than anything, that is a stub that will hopefully either work adequately when e-mail is enabled or have a better finish put on it at that time.

File: coldfront/plugins/qumulo/views/update_allocation_view.py (hunk 13 above)
Comment (ris-bmulligan) [reply coldfront/plugins/qumulo/views/update_allocation_view.py:34]: The reset ACLs UI components go with the update allocation form and view.  They are intended to not display in the create allocation form.

File: coldfront/plugins/qumulo/tasks.py (hunk 15 above)
Comment (ris-bmulligan) [reply coldfront/plugins/qumulo/tasks.py:226]: I added that because the task would throw errors about lower-level classes/objects not being "pickle"-able.  Evidently that is related to multi-threading (really async processing) or multi-processing.  I managed to get past that by making the Qumulo API object instantiation be a "check if exists, else instantiate" operation.

**Review by melonistic — APPROVED** [review]

File: coldfront/plugins/qumulo/views/update_allocation_view.py (hunk 2 above)
Comment (ris-bmulligan) [reply coldfront/plugins/qumulo/views/update_allocation_view.py:117]: It looks like start and end times ("started" and "stopped" in the admin console under successful tasks) are logged by the application.
